### PR TITLE
MOTECH-2011: Added dynamic channel functionality

### PR DESCRIPTION
**Pull request by ngraczewski**

### CODE REVIEW

**Comment by ngraczewski** [comment]
test this please

**Comment by ngraczewski** [comment]
test this please

**Comment by ngraczewski** [comment]
test this please

File: modules/scheduler/scheduler/pom.xml [hunk 1]
```diff
@@ -78,6 +78,11 @@
             <version>${project.version}</version>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.motechproject</groupId>
+            <artifactId>motech-tasks</artifactId>
+            <version>1.0-SNAPSHOT</version>
```
Comment (pgesek) [inline modules/scheduler/scheduler/pom.xml]: use project.version

File: modules/scheduler/scheduler/src/main/java/org/motechproject/scheduler/constants/SchedulerConstants.java [hunk 2]
```diff
@@ -6,6 +6,7 @@
 public final class SchedulerConstants {
     public static final String VIEW_SCHEDULER_JOBS = "hasRole('viewSchedulerJobs')";
     public static final String EVENT_TYPE_KEY_NAME = "eventType";
+    public static final String JOB_ID_KEY = "JobID";
```
Comment (pgesek) [inline modules/scheduler/scheduler/src/main/java/org/motechproject/scheduler/constants/SchedulerConstants.java]: this is already defined somewhere

File: modules/scheduler/scheduler/src/main/java/org/motechproject/scheduler/service/impl/MotechScheduledJob.java [hunk 3]
```diff
@@ -45,7 +47,8 @@ public void execute(JobExecutionContext jobExecutionContext) {
             String eventType = jobDataMap.getString(SchedulerConstants.EVENT_TYPE_KEY_NAME);
             Map<String, Object> params = jobDataMap.getWrappedMap();
             params.remove(SchedulerConstants.EVENT_TYPE_KEY_NAME);
-            params.put("JobID", jobId);
+            params.put(SchedulerConstants.JOB_ID_KEY, jobId);
+            params.put(TasksEventParser.CUSTOM_PARSER_EVENT_KEY, SchedulerEventParser.PARSER_NAME);
```
Comment (pgesek) [inline modules/scheduler/scheduler/src/main/java/org/motechproject/scheduler/service/impl/MotechScheduledJob.java]: I don't like the idea of putting this into all job executions, any way to limit this?

File: modules/scheduler/scheduler/src/main/java/org/motechproject/scheduler/tasks/SchedulerChannelProvider.java [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+import java.io.ObjectInputStream;
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.Statement;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Properties;
+
+@Service
+public class SchedulerChannelProvider implements DynamicChannelProvider {
```
Comment (pgesek) [inline modules/scheduler/scheduler/src/main/java/org/motechproject/scheduler/tasks/SchedulerChannelProvider.java]: We should mark the task dependency as optional, similar to what commcare and other modules do

File: modules/scheduler/scheduler/src/main/java/org/motechproject/scheduler/tasks/SchedulerChannelProvider.java [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+    private static final String DATA_SOURCE = "org.quartz.jobStore.dataSource";
+    private static final String JOB_NAME = "JOB_NAME";
+    private static final String JOB_DESCRIPTION = "DESCRIPTION";
+    private static final String JOB_DATA = "JOB_DATA";
+    private static final String QRTZ_JOB_DETAILS = "QRTZ_JOB_DETAILS";
+
+    @Autowired
+    private Properties sqlProperties;
+
+    public List<TriggerEvent> getTriggers(int page, int pageSize) {
```
Comment (pgesek) [inline modules/scheduler/scheduler/src/main/java/org/motechproject/scheduler/tasks/SchedulerChannelProvider.java]: This belongs in the service that deals with the scheduler database

File: modules/scheduler/scheduler/src/main/java/org/motechproject/scheduler/tasks/SchedulerChannelProvider.java [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+                            rs.getString(JOB_NAME),
+                            rs.getString(JOB_DESCRIPTION),
+                            new ArrayList<>(),
+                            dataMap.getString(SchedulerConstants.EVENT_TYPE_KEY_NAME)
+                    ));
+                }
+            }
+
+            return triggers;
+        } catch (Exception e) {
```
Comment (pgesek) [inline modules/scheduler/scheduler/src/main/java/org/motechproject/scheduler/tasks/SchedulerChannelProvider.java]: avoid catching generic Exception

File: modules/scheduler/scheduler/src/main/java/org/motechproject/scheduler/tasks/SchedulerChannelProvider.java [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+            throw new TriggerRetrievalException("Couldn't retrieve triggers for Scheduler channel", e);
+        }
+    }
+
+    @Override
+    public TriggerEvent getTrigger(String subject) {
```
Comment (pgesek) [inline modules/scheduler/scheduler/src/main/java/org/motechproject/scheduler/tasks/SchedulerChannelProvider.java]: This also belongs in the database service

File: modules/scheduler/scheduler/src/main/java/org/motechproject/scheduler/tasks/SchedulerChannelProvider.java [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+                            rs.getString(JOB_NAME),
+                            rs.getString(JOB_DESCRIPTION),
+                            new ArrayList<>(),
+                            dataMap.getString(SchedulerConstants.EVENT_TYPE_KEY_NAME)
+                    );
+                }
+            }
+
+            return null;
+        } catch (Exception e) {
```
Comment (pgesek) [inline modules/scheduler/scheduler/src/main/java/org/motechproject/scheduler/tasks/SchedulerChannelProvider.java]: avoid catching generic exception

File: modules/scheduler/scheduler/src/main/java/org/motechproject/scheduler/tasks/SchedulerChannelProvider.java [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+            throw new TriggerRetrievalException("Couldn't retrieve trigger for Scheduler channel", e);
+        }
+    }
+
+    @Override
+    public long countTriggers() {
```
Comment (pgesek) [inline modules/scheduler/scheduler/src/main/java/org/motechproject/scheduler/tasks/SchedulerChannelProvider.java]: belong in the scheduler database service

File: modules/tasks/tasks/src/main/java/org/motechproject/tasks/domain/Channel.java [hunk 10]
```diff
@@ -46,6 +49,9 @@
     @Field
     private String moduleVersion;
 
+    @Field
```
Comment (pgesek) [inline modules/tasks/tasks/src/main/java/org/motechproject/tasks/domain/Channel.java]: Whats the point of storing this in the database?

File: modules/tasks/tasks/src/test/java/org/motechproject/tasks/service/TaskTriggerHandlerTest.java [hunk 11]
```diff
@@ -321,8 +307,7 @@ public void shouldNotSendEventIfActionEventParameterCanNotBeConvertedToInteger()
         setTriggerEvent();
         setActionEvent();
 
-        when(taskService.findTrigger(TRIGGER_SUBJECT)).thenReturn(triggerEvent);
-        when(taskService.findActiveTasksForTrigger(triggerEvent)).thenReturn(tasks);
+                when(taskService.findActiveTasksForTriggerSubject(TRIGGER_SUBJECT)).thenReturn(tasks);
```
Comment (pgesek) [inline modules/tasks/tasks/src/test/java/org/motechproject/tasks/service/TaskTriggerHandlerTest.java]: what happened with formatting in this file?

**Comment by ngraczewski** [comment]
test this please

**Comment by ngraczewski** [comment]
test this please

**Comment by ngraczewski** [comment]
test this please

**Comment by pgesek** [comment]
test this please

**Comment by pgesek** [comment]
test this please

**Comment by pgesek** [comment]
test this please

**Comment by pgesek** [comment]
test this please

**Comment by ngraczewski** [comment]
test this please

**Comment by ngraczewski** [comment]
test this please

**Comment by ngraczewski** [comment]
test this please

**Comment by ngraczewski** [comment]
test this please

**Comment by ngraczewski** [comment]
test this please

**Comment by ngraczewski** [comment]
test this please

File: modules/tasks/tasks/src/main/java/org/motechproject/tasks/service/impl/TriggerEventServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+
+@Service
+public class TriggerEventServiceImpl implements TriggerEventService {
+
+    private TriggerEventsDataService triggerEventsDataService;
+    private ChannelsDataService channelsDataService;
+    private DynamicChannelLoader dynamicChannelLoader;
+
+    @Override
+    public List<TriggerEvent> getDynamicTriggers(String moduleName, int page, int pareSize) {
```
Comment (pgesek) [inline modules/tasks/tasks/src/main/java/org/motechproject/tasks/service/impl/TriggerEventServiceImpl.java]: Job id should be a field of the trigger.

File: modules/tasks/tasks/src/main/resources/webapp/partials/modals/triggersModal.html [hunk 13]
```diff
@@ -0,0 +1,77 @@
+<div class="tasks modal fade" id="triggersModal"  tabindex="-1" role="dialog">
```
Comment (pgesek) [inline modules/tasks/tasks/src/main/resources/webapp/partials/modals/triggersModal.html]: We should show how many pages there are total. We should also allow fast forward to last and first page. Would be also good to make the number an input that can be manually changed. In other words, make this more like jqgrid pagination.

File: modules/tasks/tasks/src/test/java/org/motechproject/tasks/service/TaskTriggerHandlerTest.java [hunk 14]
```diff
@@ -1387,8 +1321,7 @@ public void shouldCaptureUnrecognizedError() throws Exception {
         actionEvent.setServiceInterface("TestService");
         actionEvent.setServiceMethod("abc");
 
-        when(taskService.findTrigger(TRIGGER_SUBJECT)).thenReturn(triggerEvent);
-        when(taskService.findActiveTasksForTrigger(triggerEvent)).thenReturn(tasks);
+                when(taskService.findActiveTasksForTriggerSubject(TRIGGER_SUBJECT)).thenReturn(tasks);
```
Comment (pgesek) [inline modules/tasks/tasks/src/test/java/org/motechproject/tasks/service/TaskTriggerHandlerTest.java]: broken indentation since this spot

**Comment by ngraczewski** [comment]
test this please

**Comment by ngraczewski** [comment]
test this please

**Comment by pgesek** [comment]
Needs rebase

**Comment by ngraczewski** [comment]
test this please

**Comment by ngraczewski** [comment]
test this please

**Comment by ngraczewski** [comment]
test this please

**Comment by ngraczewski** [comment]
test this please

**Comment by ngraczewski** [comment]
test this please

**Comment by ngraczewski** [comment]
test this please

**Comment by pgesek** [comment]
test this please

**Comment by pgesek** [comment]
Needs rebase

**Comment by ngraczewski** [comment]
test this please

**Comment by pgesek** [comment]
needs merge with master

**Comment by ngraczewski** [comment]
test this please

**Comment by ngraczewski** [comment]
test this please

**Comment by ngraczewski** [comment]
test this please

**Comment by ngraczewski** [comment]
test this please

**Comment by ngraczewski** [comment]
test this please

**Comment by ngraczewski** [comment]
test this please

**Comment by ngraczewski** [comment]
test this please

**Comment by ngraczewski** [comment]
test this please

**Comment by ngraczewski** [comment]
test this please

**Comment by ngraczewski** [comment]
test this please

**Comment by ngraczewski** [comment]
test this please

**Comment by ngraczewski** [comment]
test this please

**Comment by ngraczewski** [comment]
test this please

**Comment by ngraczewski** [comment]
test this please

**Comment by ngraczewski** [comment]
test this please

**Comment by ngraczewski** [comment]
test this please

**Comment by ngraczewski** [comment]
test this please

**Pull request closed without merging**